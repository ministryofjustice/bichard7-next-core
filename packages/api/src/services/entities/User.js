"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
exports.__esModule = true;
var typeorm_1 = require("typeorm");
var BaseEntity_1 = require("./BaseEntity");
var delimitedPrefixedString_1 = require("./transformers/delimitedPrefixedString");
var featureFlagTransformer_1 = require("./transformers/featureFlagTransformer");
var User = /** @class */ (function (_super) {
    __extends(User, _super);
    function User() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.groups = [];
        return _this;
    }
    Object.defineProperty(User.prototype, "canLockTriggers", {
        get: function () {
            return this.groups.some(function (group) {
                return group === "TriggerHandler" || group === "GeneralHandler" || group === "Allocator" || group === "Supervisor";
            });
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(User.prototype, "canLockExceptions", {
        get: function () {
            return this.groups.some(function (group) {
                return group === "ExceptionHandler" || group === "GeneralHandler" || group === "Allocator" || group === "Supervisor";
            });
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(User.prototype, "isSupervisor", {
        get: function () {
            return this.groups.some(function (group) { return group === "Supervisor"; });
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(User.prototype, "visibleCases", {
        get: function () {
            return this.visibleForces.concat(this.visibleCourts);
        },
        enumerable: false,
        configurable: true
    });
    __decorate([
        (0, typeorm_1.PrimaryColumn)("varchar"),
        __metadata("design:type", String)
    ], User.prototype, "username");
    __decorate([
        (0, typeorm_1.Column)("varchar"),
        __metadata("design:type", String)
    ], User.prototype, "password");
    __decorate([
        (0, typeorm_1.Column)("varchar"),
        __metadata("design:type", String)
    ], User.prototype, "email");
    __decorate([
        (0, typeorm_1.Column)("varchar"),
        __metadata("design:type", String)
    ], User.prototype, "forenames");
    __decorate([
        (0, typeorm_1.Column)("varchar"),
        __metadata("design:type", String)
    ], User.prototype, "surname");
    __decorate([
        (0, typeorm_1.Column)({ name: "visible_forces", transformer: (0, delimitedPrefixedString_1["default"])(",", "0"), type: "varchar" }),
        __metadata("design:type", Array)
    ], User.prototype, "visibleForces");
    __decorate([
        (0, typeorm_1.Column)({ name: "visible_courts", transformer: (0, delimitedPrefixedString_1["default"])(",", "0"), type: "varchar" }),
        __metadata("design:type", Array)
    ], User.prototype, "visibleCourts");
    __decorate([
        (0, typeorm_1.Column)({ name: "feature_flags", transformer: featureFlagTransformer_1["default"], type: "jsonb" }),
        __metadata("design:type", Object)
    ], User.prototype, "featureFlags");
    User = __decorate([
        (0, typeorm_1.Entity)({ name: "users" })
    ], User);
    return User;
}(BaseEntity_1["default"]));
exports["default"] = User;
