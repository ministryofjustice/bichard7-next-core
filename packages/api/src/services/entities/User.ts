import { Column, Entity, PrimaryColumn } from "typeorm"
import BaseEntity from "../../services/entities/BaseEntity"
import delimitedPrefixedString from "../../services/entities/transformers/delimitedPrefixedString"
import featureFlagTransformer from "../../services/entities/transformers/featureFlagTransformer"
import type GroupName from "../../types/GroupName"

@Entity({ name: "users" })
export default class User extends BaseEntity {
  @PrimaryColumn("varchar")
  username!: string

  @Column("varchar")
  password!: string

  @Column("varchar")
  email!: string

  @Column("varchar")
  forenames?: string

  @Column("varchar")
  surname?: string

  @Column({ name: "visible_forces", transformer: delimitedPrefixedString(",", "0"), type: "varchar" })
  visibleForces!: string[]

  @Column({ name: "visible_courts", transformer: delimitedPrefixedString(",", "0"), type: "varchar" })
  visibleCourts!: string[]

  @Column({ name: "feature_flags", transformer: featureFlagTransformer, type: "jsonb" })
  featureFlags!: Record<string, boolean>

  groups: GroupName[] = []

  get canLockTriggers() {
    return this.groups.some(
      (group) =>
        group === "TriggerHandler" || group === "GeneralHandler" || group === "Allocator" || group === "Supervisor"
    )
  }

  get canLockExceptions() {
    return this.groups.some(
      (group) =>
        group === "ExceptionHandler" || group === "GeneralHandler" || group === "Allocator" || group === "Supervisor"
    )
  }

  get isSupervisor() {
    return this.groups.some((group) => group === "Supervisor")
  }

  get visibleCases() {
    return this.visibleForces.concat(this.visibleCourts)
  }
}
