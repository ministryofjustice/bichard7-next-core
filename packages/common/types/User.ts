export type User = {
  id: number
  username: string
  jwt_id: string
}
