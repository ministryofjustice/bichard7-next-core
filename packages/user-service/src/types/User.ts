import { UserGroupResult } from "./UserGroup"
import type FeatureFlags from "./FeatureFlags"

interface User {
  id: number
  username: string
  endorsedBy: string
  orgServes: string
  forenames: string
  surname: string
  emailAddress: string
  groups: UserGroupResult[]
  visibleCourts: string
  visibleForces: string
  excludedTriggers: string
  featureFlags?: FeatureFlags
  // We add dynamic properties to this type for form data so we need this to avoid use of unknown
  [key: string]: unknown
}

export default User
