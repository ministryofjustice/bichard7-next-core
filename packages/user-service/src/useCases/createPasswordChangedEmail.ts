import generatePasswordChangedEmail from "emails/passwordChanged"
import EmailContent from "types/EmailContent"
import User from "types/User"

export default (user: User, baseUrl: string): EmailContent => {
  const url = new URL("/login", baseUrl)

  return generatePasswordChangedEmail({ url: url.href, user })
}
