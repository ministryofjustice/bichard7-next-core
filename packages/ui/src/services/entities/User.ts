import { Column, Entity, JoinColumn, OneToMany, PrimaryColumn, Relation } from "typeorm"
import Permission from "types/Permission"
import { UserGroup } from "../../types/UserGroup"
import { userAccess } from "../../utils/userPermissions"
import delimitedString from "./transformers/delimitedString"
import jsonTransformer from "./transformers/jsonTransformer"
// eslint-disable-next-line import/no-cycle
import Note from "./Note"
// eslint-disable-next-line import/no-cycle
import SurveyFeedback from "./SurveyFeedback"

@Entity({ name: "users" })
export default class User {
  @PrimaryColumn()
  id!: number

  @PrimaryColumn()
  username!: string

  @Column()
  password!: string

  @Column()
  email!: string

  @Column()
  forenames?: string

  @Column()
  surname?: string

  @Column({ name: "visible_forces", transformer: delimitedString(","), type: "varchar" })
  visibleForces!: string[]

  @Column({ name: "visible_courts", transformer: delimitedString(","), type: "varchar" })
  visibleCourts!: string[]

  @Column({ name: "excluded_triggers", transformer: delimitedString(","), type: "varchar" })
  excludedTriggers!: string[]

  @Column({ name: "feature_flags", transformer: jsonTransformer, type: "jsonb" })
  featureFlags!: Record<string, boolean>

  @OneToMany(() => Note, (note) => note.user)
  @JoinColumn({ name: "user_id" })
  notes!: Relation<Note>[]

  @OneToMany(() => SurveyFeedback, (surveyFeedback) => surveyFeedback.user)
  @JoinColumn({ name: "user_id" })
  surveyFeedback!: Relation<User>[]

  groups: UserGroup[] = []

  get hasAccessTo(): { [key in Permission]: boolean } {
    return userAccess(this)
  }
}
