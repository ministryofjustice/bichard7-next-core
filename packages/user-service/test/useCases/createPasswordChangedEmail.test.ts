/**
 * @jest-environment node
 */

import User from "types/User"
import createPasswordChangedEmail from "useCases/createPasswordChangedEmail"

it("should generate the email content", () => {
  const user = {
    username: "Dummy username",
    emailAddress: "dummy@example.com",
    forenames: "Dummy forenames",
    surname: "Dummy surname"
  } as User

  const { subject, text, html } = createPasswordChangedEmail(user, "http://localhost:3000")

  expect(subject).toMatchSnapshot()
  expect(text).toMatchSnapshot()
  expect(html).toMatchSnapshot()
})
