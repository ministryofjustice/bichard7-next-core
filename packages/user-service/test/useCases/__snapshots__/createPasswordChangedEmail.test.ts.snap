// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should generate the email content 1`] = `"Bichard password changed"`;

exports[`should generate the email content 2`] = `
"Hi Dummy forenames Dummy surname,

Your password has been changed successfully. You can now login to your Bichard account with your new password.

http://localhost:3000/login
"
`;

exports[`should generate the email content 3`] = `"<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/><title>Bichard password changed</title></head><body><table style="background-color:#f3f2f1;margin:0;padding:30px;width:100%"><tr></tr><tr><td style="background-color:#ffffff;border:1px solid;border-color:#b1b4b6;clear:both;display:block !important;margin:0 auto;max-width:600px !important;padding:30px;vertical-align:top" valign="top" width="600"><h1 style="color:#0b0c0c;font-family:Arial, sans-serif;font-size:48px;line-height:50px;margin-bottom:50px">Bichard password changed</h1><p style="color:#0b0c0c;font-family:Arial, sans-serif;font-size:19px;line-height:25px;margin-bottom:20px">Hi Dummy forenames Dummy surname,</p><p style="color:#0b0c0c;font-family:Arial, sans-serif;font-size:19px;line-height:25px;margin-bottom:20px">Your password has been changed successfully. You can now login to your Bichard account with your new password.</p><a href="http://localhost:3000/login" style="background-color:#00703c;color:#ffffff;cursor:pointer;display:inline-block;font-family:Arial, sans-serif;font-size:19px;line-height:19px;padding:8px 10px 7px 10px;text-align:center;text-decoration:none">Login</a><p style="color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px;margin-top:50px">If youâ€™re having trouble clicking the button, copy and paste the URL below into your web browser:</p><p style="color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px;margin:0"><a href="http://localhost:3000/login" style="color:#505a5f;font-family:Arial, sans-serif;font-size:14px;line-height:16px;margin-bottom:16px">http://localhost:3000/login</a></p></td></tr><tr></tr></table></body></html>"`;
